#34186 adds shotgun_deploy and shotgun_base modules

manneohrstrom opened the pull request:
Major refactor targeting the 0.18 release. 

Highlights:
- Introduces a new `shotgun_deploy` module that can be used to bootstrap toolkit.
- Introduces a `shotgun_base` module that can holds common operations.
- Descriptor system moved out of core into a separate module for better access
- Lots of cleanup and restructuring
- Introduces a uri based descriptor syntax
- Introduces the concept of a local bundle cache where apps are stored
- Core cache hook updated to use bundle cache root by default
- Improved centralized logging, exception handling and I/O operations
- Introduces extended descriptors for file system, git and shotgun itself
- Workflows where a pipeline configuration does not exist are now possible.
- Introduces a `ToolkitManager` class that can be used to manipulate projects and configurations. Think of this as a project setup v2.0
- It is now possible to run toolkit directly from the app store with no local installation needed.

*Stuff still todo:* Address `@todos` in the code

josh-t commented on hooks/cache_location.py:
```diff
@@ -64,48 +73,33 @@ def bundle_cache(self, project_id, pipeline_configuration_id, bundle):
+        # tk-framework-shotgunutils --> fw-shotgunutils        
+        bundle_name = bundle.name
+        if bundle_name.startswith("tk-framework-"):
+            bundle_name = "fw-%s" % bundle_name[len("tk-framework-"):]
```
minor: might just be personal preference, but I feel like this reads better: 
```python
bundle_name = bundle_name.replace("tk-framework-", "fw-")
```

jfboismenu replied:
I'd say major. :) I prefer Josh's suggestion too.

manneohrstrom replied:
fixed!

josh-t commented on hooks/cache_location.py:
```diff
@@ -64,48 +73,33 @@ def bundle_cache(self, project_id, pipeline_configuration_id, bundle):
+        # tk-multi-workfiles2 --> tm-workfiles2        
+        bundle_name = bundle.name
+        if bundle_name.startswith("tk-multi-"):
+            bundle_name = "tm-%s" % bundle_name[len("tk-multi-"):]
```
minor: same comment about using `str.replace()`

manneohrstrom replied:
fixed!

josh-t commented on python/tank/deploy/tank_commands/install.py:
```diff
@@ -210,31 +206,21 @@ def _run(self, log, suppress_prompts, env_name, engine_instance_name, app_name,
-                AppDescriptor.APP, app_name
-            )
+            log.info("Connecting to the Toolkit App Store...")            
+            location = {"type": "app_store", "name": app_name, "version": "latest"}
```
so much cleaner. nice!

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor.py:
```diff
@@ -0,0 +1,286 @@
+            display_name = self.get_system_name()
+        return display_name
+
+    def is_developer(self):
```
minor: Interesting. I always thought of a `dev` descriptor representing a local development location, not a developer. I would have expected this method to be something like `is_dev()` or `is_local_dev()`.

jfboismenu replied:
Actually, I don't like the term is_developer here. Correct me if I am wrong, but the main difference between a dev descriptor and the others is that the bundle is not pulled from another source and copied into the cache, it is used directly.

I'm reluctant to use the term is_local, but maybe there's something more appropriate here?

Or maybe I am just splitting hairs?

manneohrstrom replied:
Great points. @jfboismenu - there are several modes - developer doesn't mean local, it really means developer - this is to make it clearer when to enable stuff like reload menus etc. So the dev descriptor is a "developer" mode where the "path" descriptor isn't. Similarly, there is a "immutable" flag to indicate that a descriptor will never be manually edited. @josh-t i'll change to `is_dev` - i was thinking "developer *mode*" rather than "i am the developer!". Dev is better, reflects the name of the descriptor too.

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -83,59 +81,118 @@ def __init__(self, pipeline_configuration_path):
+        self._published_file_entity_type = pipeline_config_metadata.get("published_file_entity_type", "TankPublishedFile")        
+        self._use_shotgun_path_cache = pipeline_config_metadata.get("use_shotgun_path_cache", False)
+
+        # figure out if to use the global bundle cache or the local pc 'install' cache
```
minor: "if to use" => "whether to use"

josh-t replied:
minor: For people new to Toolkit, browsing through the codebase, I think it might be better to use a spelled out "pipeline config" rather than the shortened "pc". Very minor, but I think it would help with readability in the long run.

manneohrstrom replied:
Done - renamed all pc in comments to pipeline config.

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -83,59 +81,118 @@ def __init__(self, pipeline_configuration_path):
+        # get current settings
+        curr_settings = self._get_metadata()
+        
+        # add path cache setting
```
This comment seems to know in advance what the specific updates are? Is that right?

manneohrstrom replied:
old comment! clarified!

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -83,59 +81,118 @@ def __init__(self, pipeline_configuration_path):
+    
+        return data
+    
+    def _update_pipeline_config(self, updates):
```
minor: Maybe this should be called `_update_metadata()` for consistency with `_get_metadata()`.

manneohrstrom replied:
fixed

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -83,59 +81,118 @@ def __init__(self, pipeline_configuration_path):
+            raise TankError("Configuration metadata file '%s' missing! "
+                            "Please contact support." % cfg_yml)
+    
+        fh = open(cfg_yml, "rt")
```
Should the `open()` be inside the try?

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -236,35 +298,15 @@ def is_localized(self):
-        Returns the shotgun id for the project associated with this PC. 
-        May connect to Shotgun to retrieve this.
+        Returns the shotgun id for the project associated with this PC.
+        Can return None if the pipeline configuratoin represents the site and not a project.
```
typo: "configuratoin"

... still better than PC!!! :P

manneohrstrom replied:
fxed

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -559,54 +531,101 @@ def get_install_location(self):
+        if location_dict.get("type") == "dev":
+
+            # platform specific location support
+            platform_key = {"linux2": "linux_path", "darwin": "mac_path", "win32": "windows_path"}[sys.platform]
```
This syntax is not my favorite. Looks like this lookup is used quite a bit throughout the codebase. Perhaps this logic could be promoted to a shared location?
![screen shot 2016-02-23 at 8 40 17 pm](https://cloud.githubusercontent.com/assets/2604646/13272930/bc250f3a-da6d-11e5-9ef7-71639fcb2211.png)

jfboismenu replied:
I second that.

jfboismenu replied:
There's another pattern that I don't like also that I find everywhere, the platform detection code using sys.platform and doing a partial match on linux

Why don't be introduce
is_macos (is_osx won't be valid when they go to 11)
is_windows
is_linux in shotgun_base?

manneohrstrom replied:
Nice! I have added a method to `tank_base` for this and cleaned up all places where we had a fixed dict to jump from os.system -> "windows_path"/"linux_path" etc.

manneohrstrom replied:
@jfboismenu your comment doesn't feel directly related to this CR, so I'll skip that for now. Enough changes already as it is :) Maybe it's cleaner to introduce wrapper methods, but if i understand you correctly, we are not adding any logic of our own here, right? `is_linux()` is equal to `sys.platform == "linux2"`.

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -559,54 +531,101 @@ def get_install_location(self):
+
+        :param descriptor_type: Descriptor type (APP, ENGINE, etc)
+        :param location:        Location dictionary
+        :param latest:          Return latest available version of descriptor
```
no longer a param it seems.

manneohrstrom replied:
fixed

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -559,54 +531,101 @@ def get_install_location(self):
-        Note! This method has been deprecated and will be removed at 
-        some point in the future. Please do not use it.
+        :param location: Location dictionary describing the app source location
+        :param latest: Return latest available version of descriptor
```
remove?

manneohrstrom replied:
fixed

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -559,54 +531,101 @@ def get_install_location(self):
         
-        :returns: path string        
+        :param location: Location dictionary describing the engine source location
+        :param latest: Return latest available version of descriptor
```
remove?

manneohrstrom replied:
fixed

josh-t commented on python/tank/platform/environment.py:
```diff
@@ -290,9 +289,7 @@ def get_framework_descriptor(self, framework_name):
-        d = descriptor.get_from_location(descriptor.AppDescriptor.FRAMEWORK,
-                                         self.__pipeline_config,
-                                         location_dict)
+        d = self.__pipeline_config.get_framework_descriptor(location_dict)
```
minor: For these methods, maybe instead of `d = self.` just `return self.`? Would avoid use of nondescript `d` variable.

manneohrstrom replied:
yes! weird code. fixed.

josh-t commented on python/tank_vendor/shotgun_authentication/session_cache.py:
```diff
@@ -29,9 +29,11 @@
-logger = logging.getLogger("sg_auth.session")
+from ..shotgun_base import get_cache_root, get_site_cache_root
+
+logger = get_logger("session")
```
I'm curious about the removal of the prefixed logger names. I realize they weren't consistent, but wondering why no prefix anymore. Perhaps a client might want to filter all shotgun/toolkit related logging?

josh-t replied:
doh. nm. I see what's happening now. :+1:

josh-t commented on python/tank_vendor/shotgun_base/log.py:
```diff
@@ -0,0 +1,50 @@
+    """
+    Returns a official sgtk root logger.
+
+    If the name parameter is left as None, the root
```
these lines seem prematurely wrapped.

josh-t commented on python/tank_vendor/shotgun_base/paths.py:
```diff
@@ -0,0 +1,124 @@
+    """
+
+    # get site only; https://www.foo.com:8080 -> www.foo.com
+    base_url = urlparse.urlparse(site_url)[1].split(":")[0]
```
very minor: use `.netloc` instead of `[1]`. ever so slightly more readable.

manneohrstrom replied:
changed

josh-t commented on python/tank_vendor/shotgun_base/paths.py:
```diff
@@ -0,0 +1,124 @@
+    # shotgun.internal.int     -> shotgun.internal.int
+    #
+    if base_url.endswith("shotgunstudio.com"):
+        base_url = base_url[:-len(".shotgunstudio.com")]
```
minor: for readability:
```python
base_url = base_url.replace(".shotgunstudio.com", "")
```

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+        - Sets up basic folder structure for a config
+        - Copies the configuration into place.
+
+        :returns: True if a scaffold was created, False if a complete or
```
Maybe I'm reading this wrong, but the way the code looks, it looks like `config_exists` returns `True` if the path already existed, `False` otherwise.

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+        Creates all the necessary files on disk for a basic config scaffold.
+
+        - Sets up basic folder structure for a config
+        - Copies the configuration into place.
```
I don't see where the config is copied into place.

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+            project_name = "Site"
+            pipeline_config_name = constants.UNMANAGED_PIPELINE_CONFIG_NAME
+
+        data = {
```
minor: `return {`

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+            "published_file_entity_type": "PublishedFile",
+            "use_global_bundle_cache": True,
+            "bundle_cache_fallback_roots": self._bundle_cache_fallback_paths,
+            "use_shotgun_path_cache": True}
```
minor: `}` on next line aligned to match opening `{`

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+        """
+        config_path = self.get_path()
+
+        # write a file location file for our new setup
```
minor: "a file location file" => "the install location file"

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor.py:
```diff
@@ -0,0 +1,286 @@
+import copy
+
+from . import paths
+from .descriptor_io import create_io_descriptor
```
minor: `descriptor_io` vs. `io_descriptor`. Wondering if there should be ore consistency by calling the file `io_descriptor` since that seems to be how these are referred to (`IODescriptor*` classes etc) throughout the code. I see the form is mirrored with `bundle`, `config`, and `core` descriptors too. Just something to consider.

manneohrstrom replied:
i don't have a strong opinion. TBH, I was hoping CR would suggest an even better name than `DescriptorIO`. Changed as suggested!

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+            # not sure what version this is.
+            return self.LOCAL_CFG_INVALID
+
+        fh = open(config_info_file, "rt")
```
shoudl this be in the `try` perhaps?

josh-t commented on python/tank/deploy/tank_commands/setup_project_core.py:
```diff
@@ -277,44 +278,44 @@ def _project_setup_internal(log, sg, sg_app_store, sg_app_store_script_user, set
-    
+
     # write the record to disk
     pipe_config_sg_id_path = os.path.join(config_location_curr_os, "config", "core", "pipeline_configuration.yml")
```
file name is in `constants` now. use that instead?

manneohrstrom replied:
i wanted to make as few changes to the old setup code as possible, this is why this is not using the constant. Updated now!

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -83,59 +81,118 @@ def __init__(self, pipeline_configuration_path):
+        # now read in the pipeline_configuration.yml file
+        cfg_yml = os.path.join(self.get_config_location(), 
+                               "core", 
+                               "pipeline_configuration.yml")
```
use filename from constants?

manneohrstrom replied:
fixed!

josh-t commented on python/tank/pipelineconfig.py:
```diff
@@ -83,59 +81,118 @@ def __init__(self, pipeline_configuration_path):
+        # write the record to disk
+        pipe_config_sg_id_path = os.path.join(self.get_config_location(), 
+                                              "core", 
+                                              "pipeline_configuration.yml")        
```
use filename from constants?

manneohrstrom replied:
fixed!

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+
+        # first check if there is any config at all
+        # probe for shotgun.yml connection params file
+        if not os.path.exists(os.path.join(config_root, "config", "core", "shotgun.yml")):
```
use filename from constants?

manneohrstrom replied:
fixed :)

josh-t commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,965 @@
+                ["code", "project.Project.tank_name"]
+            )
+
+            project_name = sg_data["project.Project.tank_name"] or "Unnamed"
```
perhaps unnamed project name should come from constants?

manneohrstrom replied:
fixed

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/appstore.py:
```diff
@@ -0,0 +1,656 @@
+    
+            # now get the version
+            version = sg.find_one(
+                version_entity,
```
`version` all by itself is *always* confusing in SG. :(

The `version` variable might be better as `version_entity` `bundle_version_entity` or `bundle_version_entity_name`.

The current `version_entity` variable might be better as `version_entity_type` or `bundle_version_entity_type`. 

(assuming I'm reading the code correctly of course)

manneohrstrom replied:
yep agree - this is old code that still needs some love :) will try to clean up

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/appstore.py:
```diff
@@ -0,0 +1,656 @@
+                  is an sg entity dictionary (keys type/id) corresponding to to the user used
+                  to connect to the app store.
+        """
+        # maintain a cache for performance
```
comment is oddly wrapped.

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/base.py:
```diff
@@ -0,0 +1,548 @@
+
+        # shorthand for constants
+        uri_sep = constants.LOCATOR_URI_SEPARATOR
+        uri_safe_sep = constants.ESCAPED_LOCATOR_URI_SEPARATOR
```
why not `urllib.quote()` the chunks?

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/factory.py:
```diff
@@ -0,0 +1,170 @@
+
+    :param sg: Shotgun connection to associated site
+    :param descriptor_type: Either AppDescriptor.APP, CORE, ENGINE or FRAMEWORK
+    :param location: A std location dictionary dictionary or string
```
cool! so you could use the uri syntax in the config?

manneohrstrom replied:
yup :)

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/dev.py:
```diff
@@ -0,0 +1,151 @@
+    String urls are on the following form::
+
+        sgtk:dev:[name]:local_path
+        sgtk:dev3:[name]:win_path:linux_path:mac_path
```
`dev3`?

manneohrstrom replied:
this is all TBD

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/dev.py:
```diff
@@ -0,0 +1,151 @@
+            location_dict = cls._explode_uri(
+                uri,
+                "dev3",
+                ["name", "windows_path", "linux_path", "mac_path"]
```
random thoughts... feel free to ignore: 

The new uri stuff is really cool. I like that you can identify a location via string. I had a couple of thoughts/questions just for the sake of conversation...

1. Crazy thought... Would it make any sense to leverage the templating system to create a spec for each uri type? Maybe the templating system is strictly path based, but I was wondering if there would be any benefit either in terms of customization or future proofing to externalize the location uris specs: 
```yaml
def_template: "sgtk:dev:{name}:{path}"
```
2. Assuming 1 is just ridiculous, I'm curious why not use a standard URI format with a custom scheme such as `sgtk:///my-app?path=/tmp/foo/bar`. That's not very pretty, but would allow for standard parsing, encoding, optional params, etc. Perhaps other benefits down the road somewhere?

manneohrstrom replied:
great! current format is placeholder, i didn't even try to come up with someting sensbible, was more trying to get the structure and high level concepts right for bootstrap. @robblau suggested `sgtk:descriptor:app_store?name=foo&version=bar` which I really like. I'll put your comments on our hot topics doc for discussion :)

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/factory.py:
```diff
@@ -0,0 +1,170 @@
+    else:
+        location_dict = location
+
+    if location_dict.get("type") == "app_store":
```
I can imagine, once this takes off, studios wanting to implement their own custom descriptors. Perhaps we should consider a registration system?

jfboismenu replied:
Let's do a meta-baseclass that automatically registers the derived class automatically when you derive from it! :)

manneohrstrom replied:
yes yes!, we should and this is something we have talked about for a while... the way i see it, there will always be a need for some "base" collection of descriptors, so what we do now isn't going to be undone by making it more customizable in future. The exact implementation details need to be discussed I think - again i'll whack it on hot topics. We may want to do discuss future use cases before committing to a custom format too - i am thinking in about the idea of using descriptor URIs to include config files that have been brainstormed a little bit.

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/factory.py:
```diff
@@ -0,0 +1,170 @@
+        raise ShotgunDeployError("Invalid uri %s" % location_uri)
+    descriptor_type = urllib.unquote(chunks[1])
+
+    if descriptor_type == "app_store":
```
Registration system would be nice here as well. Each descriptor type I would image is registered with a `type` to match against and you could just ask the registry for the class with the matching type.

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/factory.py:
```diff
@@ -0,0 +1,170 @@
+        raise ShotgunDeployError("Invalid location dictionary %s" % location_dict)
+    descriptor_type = location_dict["type"]
+
+    if descriptor_type == "app_store":
```
same comment regarding a registry.

manneohrstrom replied:
let's KISS it for now and brainstorm how to make it customizable.

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/git.py:
```diff
@@ -0,0 +1,337 @@
+
+log = util.get_shotgun_deploy_logger()
+
+class IODescriptorGit(IODescriptorBase):
```
You mentioned that you wished the "git" type was called "git_branch". Would it be worth changing this to `IODescriptorGitTag` and making "git_branch" a valid type as well? You'd still need "git" for backward compatibility, but maybe that could be handled with a subclass that just sets a class variable to assign the type name or something like that.

josh-t replied:
Wondering if it would make sense to have a git descriptor base class since there is shared functionality between the 2 existing classes. I think if you externalized the `required` and `optional` fields, you could reduce the code duplication even more. Just a thought.

manneohrstrom replied:
it's `git` due to backwards compatibility. But i'll shift it round with a base class. That sounds cool!

manneohrstrom replied:
shifted it round - not a massive change. Feel a little reluctant about this - never been a big fan of deep inheritance hierarchies, but leaving it for now.

josh-t commented on python/tank_vendor/shotgun_deploy/descriptor_io/git.py:
```diff
@@ -0,0 +1,337 @@
+
+    Uris are on the form:
+
+        sgtk:git:path/to/git/repo:v12.3.4
```
should this be `sgtk:git:/path/to/git/repo:v12.3.4`? Note the `/` before `path`.

josh-t commented:
Some of the files don't show up in the diff since the PR is so large. I *think* the files that aren't displayed are all new, so I'm just going through them individually...

josh-t commented:
https://github.com/shotgunsoftware/tk-core/blob/34186_plugin_spike/python/tank_vendor/shotgun_deploy/util.py#L116

Do we need to make sure this works in 2.5?

jfboismenu commented on hooks/cache_location.py:
```diff
@@ -36,12 +36,21 @@ def path_cache(self, project_id, pipeline_configuration_id):
         is stored on a local machine, separate for each user.  
-        
+
+        Note! In the case of the site configuration, project id will be set to None.
```
Is it possible these are 0 now actually?

manneohrstrom replied:
not sure what use case you refer to!

jfboismenu commented on python/tank/deploy/descriptor.py:
```diff
@@ -12,551 +12,43 @@
 
-    :param app_or_engine: Either AppDescriptor.APP AppDescriptor.ENGINE or FRAMEWORK
-    :param pc_path: Path to the root of the pipeline configuration
+    LEGACY - Use create_descriptor instead.
```
Use shotgun_deploy.create_descriptor instead.

jfboismenu replied:
We will have to make sure the current desktop startup still works with this version of core. Unfortunately we used some private APIs from the sgtk.deploy module in the desktop startup to auto-update the desktop statup.

manneohrstrom replied:
Yep yep - this is *exactly* why i kept this. Generally I have been trying to be super mindful about this, but as far as my grep-fu would allow, this is the only place i found where we were calling directly, so not too bad, in the grand scheme of things! I'll add a comment about it.

jfboismenu commented on python/tank/deploy/tank_commands/desktop_migration.py:
```diff
@@ -51,6 +51,11 @@ def run_interactive(self, log, args):
 
         pc = self.tk.pipeline_configuration
 
+        if pc.is_managed():
```
Why? Because it should actually.

manneohrstrom replied:
cool, if you say so! Maybe i was going over this too quickly! :) I'll take it out, makes sense!

jfboismenu commented on python/tank/deploy/tank_commands/move_pc.py:
```diff
@@ -97,7 +97,12 @@ def _copy_folder(self, log, level, src, dst):
     def run_interactive(self, log, args):
-        
+
+        if self.tk.pipeline_configuration.is_unmanaged():
```
Instead of having to update all commands to check for the flag, maybe there should be a property on each command that says if the can operate of managed or unmanaged and depending on the config we are using they show up or not?

manneohrstrom replied:
I like the idea, but a lot of the commands that we are making changes are now are effectively becoming legacy commands that are in some way replaced by bootstrap equivalents - setup, localize, share core, push, clone, move... all are now part of a new framework. We have some tickets to go over the tank commands and make sure they are all making sense and using the new code - I think this is a good opportunity to dive a bit deeper here too and maybe discuss this. For now, i just added the obvious little catches..

jfboismenu commented on python/tank/deploy/tank_commands/setup_project_params.py:
```diff
@@ -285,86 +285,86 @@ def get_configuration_shotgun_info(self):
         Sets up the associated template configuration. Copies files.
-        
-        :param target_path: Location where the config should be set up. 
+
```
See, that's what I meant yesterday by let's run a script and be done with it. :(

jfboismenu commented on python/tank/pipelineconfig.py:
```diff
@@ -178,27 +235,23 @@ def get_all_os_paths(self):
+        auto path status.
+
+        January 2016:
+        DEPRECATED - DO NOT USE! At some stage this wil be removed.
```
We have to keep in mind that people who are using the old desktop 1.0 who might not be upgrading to the latest desktop will still require this method to exist. We will have to push an update for the desktop startup that doesn't use this method and only several months after will it be reasonsable to remove this method.

By the way, I think we should log warnings in the console for deprecated calls. What do you think?

manneohrstrom replied:
Deprecated calls sounds great! Part of further logging work I guess! Re deprecation - I am guessing we won't take this out for at least a year. It's a warning saying 'don't write *new* code that calls this.' So I think we are good for the time being. Backwards compatbility is key!

jfboismenu commented on python/tank/pipelineconfig.py:
```diff
@@ -178,27 +235,23 @@ def get_all_os_paths(self):
+
         :returns: boolean indicating auto path state
         """
+        if self.is_unmanaged():
```
A comment explaining why would help here. I thought this was wrong at first. Something like:

"Auto path configuration auto updates the configuration on a per bundle basis, while unmanaged pipeline configurations update on a per configuration basis."

manneohrstrom replied:
done

jfboismenu commented on python/tank/pipelineconfig.py:
```diff
@@ -225,7 +278,16 @@ def _is_empty(d):
+    def is_unmanaged(self):
+        """
+        Returns true if the configuration is unmanaged, e.g. it does not have a
+        corresponding path cache in Shotgun.
```
corresponding "pipeline configuration"

manneohrstrom replied:
fixed

jfboismenu commented on python/tank/pipelineconfig.py:
```diff
@@ -310,6 +335,7 @@ def convert_to_site_config(self):
         Converts the pipeline configuration into the site configuration.
         """
         self._update_pipeline_config({"project_id": None})
+        self._project_id = None
```
What is going on here? Doesn't update_pipeline_config already set project_id to None?

manneohrstrom replied:
yep! now gone :)

jfboismenu commented on python/tank/pipelineconfig.py:
```diff
@@ -559,54 +531,101 @@ def get_install_location(self):
+        :param descriptor_type: Descriptor type (APP, ENGINE, etc)
+        :param location:        Location dictionary
+        :param latest:          Return latest available version of descriptor
+        :returns:               Descriptor object
```
weird padding between :returns: and the type that doesn't seem to be there in other docstrings

manneohrstrom replied:
will fix

jfboismenu commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,983 @@
+        """
+        return self._descriptor
+
+    def get_path(self, platform=sys.platform):
```
I wish we could leave Python 2.5 behind and use the [abc](https://docs.python.org/2/library/abc.html) module.

jfboismenu commented on python/tank_vendor/shotgun_deploy/configuration.py:
```diff
@@ -0,0 +1,983 @@
+
+        log.debug("Creating install_location file...")
+
+        # platforms other than the current OS will not be supposed
```
supported

manneohrstrom replied:
lol fixed.

jfboismenu commented on python/tank_vendor/shotgun_deploy/descriptor.py:
```diff
@@ -0,0 +1,296 @@
+    :param descriptor_type: Either AppDescriptor.APP, CORE, ENGINE or FRAMEWORK
+    :param location: A std location dictionary dictionary or string
+    :param bundle_cache_root: Optional override for root path to where downloaded apps are cached
+    :param fallback_roots: List of immutable fallback cache locations where
```
"Optional list"

manneohrstrom replied:
fixed

jfboismenu commented on python/tank_vendor/shotgun_deploy/descriptor.py:
```diff
@@ -0,0 +1,296 @@
+        """
+        meta = self._io_descriptor.get_manifest()
+        support_url = meta.get("support_url")
+        if support_url is None:
```
Not sure this is a good idea. Imagine a world where clients are hosting their configs on Github for other people to use. If they forget to fill in the field, we'll be the ones reported as doing support for it. I'd rather say let's fix the manifests of our stuff.

manneohrstrom replied:
this is super old code. I was merely converting it to the new module.

jfboismenu commented on python/tank_vendor/shotgun_deploy/descriptor_io/appstore.py:
```diff
@@ -0,0 +1,656 @@
+
+        return metadata
+
+    def _get_cache_paths(self):
```
I was looking at that code yesterday and I was wondering if it shouldn't return only paths that exist?

manneohrstrom replied:
I/O operations are comparably expensive so I am specifically *not* doing that. This method merely computes things. Will add comment to underline this.

jfboismenu commented on python/tank_vendor/shotgun_deploy/descriptor_io/appstore.py:
```diff
@@ -0,0 +1,656 @@
+        # sgtk:app_store:tk-core:v12.3.4
+
+        # explode into dictionary
+        location_dict = cls._explode_uri(uri, "app_store", ["name", "version"])
```
Shouldn't we have a property on the class that has the same of the type? Then some of this could move to a base class. The get_cache_paths could also move into a base class for versioned descriptors

josh-t commented:
Toolkit Manager API thoughts:

* do we need the `set_*` methods?
    * could these be exposed as properties?
    * any reason not to allow them as optional args to constructor?

* The method is called `set_bundle_cache_search_path()` but internally called `bundle_cache_fallback_paths`
    * Using `fallback` seems more descriptive of what it actually does
    * Suggest: `set_bundle_cache_fallback_paths()`

* `set_base_configuration` - maybe whould use same `fallback` terminology here 
    * Suggest: `set_fallback_configuration`?

* would there be any value to also having `bootstrap_app` method? 
    * possibly for things like stand-alone publisher?

* `upload_configuration()` and `create_managed_configuration()` require Toolkit to be turned on for site.
    * perhaps those should error out early on with descriptive message about how to turn tk on?
    * perhaps those methods should live under pipeline config api?

* `bootstrap_engine()`: maybe supplied entity already has project data?
    * could avoid a query?

* Overall the `ToolkitManager` API thoughts
    * The `bootstrap_sgtk()` method is super convenient, but I kinda want the control of what/where 
        it boots from to be outside of that method
    * Maybe it would be more powerful to just supply `bootstrap_sgtk` with a `config` object, and do the `resolve` externally
    * The internal members of the class are storing state that already lives in a config right?
    * Could any of these methods be moved over to the `Configuration` object?
    * I realize this is a convenience class of sorts, but seems like separating the config resolving from bootstrapping might clean things up. 
    * You've pondered this way more than I have so feel free to tell ignore this!

manneohrstrom commented:
Here are comments to @josh-t 's great feedback:

> do we need the set_* methods?

I wanted to wait with adding properties until things had settled. Now done!

> The method is called set_bundle_cache_search_path() - Suggest: set_bundle_cache_fallback_paths()

yes, I agree it's clearer! Changed

> set_base_configuration - maybe whould use same fallback terminology here?

The config plays the role both as fallback and as source for various operations. For uploading, it doesn't make sense to call it fallback. IMHO

> would there be any value to also having bootstrap_app method?

We can always add it later on if/when a need arises!

> upload_configuration() and create_managed_configuration() require Toolkit to be turned on for site.

I have added checks. For the initial release i suspect we will remove these or comment out.

> bootstrap_engine(): maybe supplied entity already has project data?

Good idea! Fixed.

> Overall the ToolkitManager API thoughts

I tried to keep the API as light weight as possible, so that we can shuffle things round internally if needed. We can always expose stuff later on. I have added this to the hot topics doc!

manneohrstrom commented:
Yay! This was merged into `develop/0.18`. Closing.
